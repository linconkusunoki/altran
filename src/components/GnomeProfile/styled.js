import styled from 'styled-components'

export const Wrapper = styled.div``

export const PersonalInfo = styled.div`
  @media (min-width: 768px) {
    display: flex;
    align-items: center;
  }
`

export const WrapperInfo = styled.div`
  margin: 24px 0;

  @media (min-width: 768px) {
    padding: 24px;
    margin: 0;
  }
`

export const Info = styled.div`
  display: flex;
  align-items: center;
  margin: 0;
  font-size: 20px;

  div {
    min-width: 100px;
    text-align: right;
    margin-right: 5px;
  }

  span {
    font-weight: 700;
    color: rgba(14, 30, 37, 0.84);
  }
`

export const Name = styled.h3`
  color: rgba(14, 30, 37, 0.87);
  font-size: 24px;
  line-height: 30px;
  margin: 0;

  @media (min-width: 768px) {
    flex-basis: 50%;
  }
`

export const Detail = styled.fieldset`
  border: 2px solid #ddd;
  border-radius: 6px;
  margin: 24px 0;
`
