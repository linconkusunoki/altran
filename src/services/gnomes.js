import api from './api'

const get = async id => {
  let res

  try {
    res = await api.get('/rrafols/mobile_test/master/data.json')
  } catch (error) {
    return error
  }

  if (id) {
    const gnome = res.data.Brastlewark.find(gnome => gnome.id === parseInt(id))
    return gnome
  }

  return res.data.Brastlewark
}

export default {
  get,
}
