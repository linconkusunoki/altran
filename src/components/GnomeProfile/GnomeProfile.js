import React from 'react'

import gnomesService from '../../services/gnomes'
import useFetch from '../../hooks/useFetch'
import Avatar from '../Avatar'
import HairColor from '../HairColor'
import Gender from '../Gender'

import * as S from './styled'

const GnomeProfile = ({ id }) => {
  const { isLoading, error, response: gnome } = useFetch(gnomesService.get, id)

  if (isLoading) {
    return <p>Loading gnome...</p>
  }

  if (error) {
    return <p>Error while loading gnome!</p>
  }

  return (
    <S.Wrapper>
      <S.BackLink to="/">Home</S.BackLink> / Gnome Profile
      <S.PersonalInfo>
        <Avatar thumbnail={gnome.thumbnail} variant="profile" />
        <S.WrapperInfo>
          <S.Title>Gnome Profile</S.Title>
          <S.Name>{gnome.name}</S.Name>
          <S.Info>
            <div>Age:</div>
            <span>{gnome.age}</span>
          </S.Info>
          <S.Info>
            <div>Weight:</div>
            <span>{gnome.weight.toFixed(2)}</span>
          </S.Info>
          <S.Info>
            <div>Height:</div>
            <span>{gnome.height.toFixed(2)}</span>
          </S.Info>
          <S.Info>
            <div>Hair color:</div>
            <HairColor color={gnome.hair_color} />
          </S.Info>
          <S.Info>
            <div>Gender:</div>
            <Gender color={gnome.hair_color} />
          </S.Info>
        </S.WrapperInfo>
      </S.PersonalInfo>
      <S.Detail>
        <legend>Professions</legend>
        <ul>
          {gnome.professions.map(profession => (
            <li key={profession}>{profession}</li>
          ))}
          {gnome.professions.length === 0 && 'Time is money friend'}
        </ul>
      </S.Detail>
      <S.Detail>
        <legend>Friends</legend>
        <ul>
          {gnome.friends.map(friend => (
            <li key={friend}>{friend}</li>
          ))}
          {gnome.friends.length === 0 && 'Stop working and go make new friends'}
        </ul>
      </S.Detail>
    </S.Wrapper>
  )
}

export default GnomeProfile
