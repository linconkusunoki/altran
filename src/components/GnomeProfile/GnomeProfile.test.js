import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { render, wait } from '@testing-library/react'
import GnomeProfile from './'
import gnomes from '../../services/gnomes'

jest.mock('../../services/gnomes')

const gnome = {
  age: 306,
  friends: ['Cogwitz Chillwidget', 'Tinadette Chillbuster'],
  hair_color: 'Pink',
  height: 107.75835,
  id: 0,
  name: 'Tobus Quickwhistle',
  professions: [
    'Metalworker',
    'Woodcarver',
    'Stonecarver',
    ' Tinker',
    'Tailor',
    'Potter',
  ],
  thumbnail:
    'http://www.publicdomainpictures.net/pictures/10000/nahled/thinking-monkey-11282237747K8xB.jpg',
  weight: 39.065952,
}

it('should render correctly', async () => {
  gnomes.get.mockResolvedValueOnce(gnome)

  const { getByText, container } = render(
    <MemoryRouter>
      <GnomeProfile id="0" />
    </MemoryRouter>,
  )

  expect(getByText('Loading gnome...'))
  await wait(() => expect(getByText('Tobus Quickwhistle')).toBeTruthy())
  expect(gnomes.get).toHaveBeenCalled()
  expect(container.firstChild).toMatchSnapshot()
})

it('should render without professions', async () => {
  const gnomeWithoutJob = { ...gnome, professions: [] }
  gnomes.get.mockResolvedValueOnce(gnomeWithoutJob)

  const { getByText } = render(
    <MemoryRouter>
      <GnomeProfile id="0" />
    </MemoryRouter>,
  )

  expect(getByText('Loading gnome...'))
  await wait(() => expect(getByText('Time is money friend')).toBeTruthy())
  expect(gnomes.get).toHaveBeenCalled()
})

it('should render without friends', async () => {
  const gnomeWithoutJob = { ...gnome, friends: [] }
  gnomes.get.mockResolvedValueOnce(gnomeWithoutJob)

  const { getByText } = render(
    <MemoryRouter>
      <GnomeProfile id="0" />
    </MemoryRouter>,
  )

  expect(getByText('Loading gnome...'))
  await wait(() =>
    expect(getByText('Stop working and go make new friends')).toBeTruthy(),
  )
  expect(gnomes.get).toHaveBeenCalled()
})

it('should render an error message', async () => {
  gnomes.get.mockRejectedValueOnce(new Error('Async error'))

  const { getByText, container } = render(
    <MemoryRouter>
      <GnomeProfile id="?" />
    </MemoryRouter>,
  )

  expect(getByText('Loading gnome...'))
  await wait(() => expect(getByText('Error while loading gnome!')).toBeTruthy())
  expect(gnomes.get).toHaveBeenCalled()
  expect(container.firstChild).toMatchSnapshot()
})
