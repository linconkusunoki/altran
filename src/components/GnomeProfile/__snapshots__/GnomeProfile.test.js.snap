// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render an error message 1`] = `
<p>
  Error while loading gnome!
</p>
`;

exports[`should render correctly 1`] = `
<div
  class="sc-bxivhb bUZGRr"
>
  <a
    class="sc-bZQynM cykBRJ"
    href="/"
  >
    Home
  </a>
   / Gnome Profile
  <div
    class="sc-ifAKCX iGTNyI"
  >
    <figure
      class="sc-bdVaJa dTbTjr"
    />
    <div
      class="sc-EHOje jEWxeX"
    >
      <h3
        class="sc-htoDjs gCyVGZ"
      >
        Gnome Profile
      </h3>
      <h3
        class="sc-dnqmqq ikpkhV"
      >
        Tobus Quickwhistle
      </h3>
      <div
        class="sc-gzVnrw eKlHpm"
      >
        <div>
          Age:
        </div>
        <span>
          306
        </span>
      </div>
      <div
        class="sc-gzVnrw eKlHpm"
      >
        <div>
          Weight:
        </div>
        <span>
          39.07
        </span>
      </div>
      <div
        class="sc-gzVnrw eKlHpm"
      >
        <div>
          Height:
        </div>
        <span>
          107.76
        </span>
      </div>
      <div
        class="sc-gzVnrw eKlHpm"
      >
        <div>
          Hair color:
        </div>
        <span
          class="sc-htpNat hWCfcZ"
          color="Pink"
        />
      </div>
      <div
        class="sc-gzVnrw eKlHpm"
      >
        <div>
          Gender:
        </div>
        <span
          aria-label="female"
          data-testid="female"
          role="img"
        >
          ðŸ‘©
        </span>
      </div>
    </div>
  </div>
  <fieldset
    class="sc-iwsKbI hZKZoG"
  >
    <legend>
      Professions
    </legend>
    <ul>
      <li>
        Metalworker
      </li>
      <li>
        Woodcarver
      </li>
      <li>
        Stonecarver
      </li>
      <li>
         Tinker
      </li>
      <li>
        Tailor
      </li>
      <li>
        Potter
      </li>
    </ul>
  </fieldset>
  <fieldset
    class="sc-iwsKbI hZKZoG"
  >
    <legend>
      Friends
    </legend>
    <ul>
      <li>
        Cogwitz Chillwidget
      </li>
      <li>
        Tinadette Chillbuster
      </li>
    </ul>
  </fieldset>
</div>
`;
